Public/Src/Cache/ContentStore/Library/Stores/FileSystemContentStoreInternalChecker.cs

```diff
@@ -0,0 +1,465 @@
+            SelfCheckState selfCheckState,
+            SelfCheckStatus status)
+        {
+            // Self checking procedure validates that in-memory content directory
+            // is valid in respect to the state on disk.
+            // Namely, it checks that the hashes for all the files and their size are correct.
+            var stopwatch = Stopwatch.StartNew();
+
+            // Enumerating files from disk instead of looking them up from content directory.
+            // This is done due to simplicity (we don't have to worry about replicas) and because an additional IO cost is negligible compared to the cost of rehashing.
+            var contentHashes = _contentStoreInternal.ReadContentHashesFromDisk(context).OrderBy(tpl => tpl.hash).ToList();
+            _tracer.Debug(context, $"SelfCheck: Enumerated {contentHashes.Count} entries from disk by {stopwatch.ElapsedMilliseconds}ms.");
+
+            stopwatch.Restart();
+
+            // Trying to restore the index of a hash that we processed before.
+            int index = 0;
+            if (status == SelfCheckStatus.InProgress)
+            {
+                index = findNextIndexToProcess(selfCheckState.LastPosition.Value);
+                _tracer.Debug(context, $"SelfCheck: skipping {index} elements based on previous state '{selfCheckState.ToParseableString()}'.");
+            }
+            else
+            {
+                string statusAsString = status == SelfCheckStatus.Force ? "the epoch has changed" : "a self check is out of date";
+                _tracer.Debug(
+                    context,
+                    $"SelfCheck: starting self check for the entire content directory because {statusAsString}. Previous state '{selfCheckState.ToParseableString()}'.");
+            }
+
+            // Task for tracking progress.
+            Task progressReportingTask = Task.Delay(_settings.SelfCheckProgressReportingInterval);
+
+            int invalidEntries = 0;
+            int processedEntries = 0;
+
+            for (; index < contentHashes.Count; index++)
+            {
+                if (context.Token.IsCancellationRequested)
+                {
+                    _tracer.Debug(context, "SelfCheck: Exiting self check because cancellation was requested.");
+                    break;
+                }
+
+                var hashInfo = contentHashes[index];
+                processedEntries++;
+
+                var (isValid, error) = await ValidateFileAsync(context, hashInfo.hash, hashInfo.fileInfo);
+                if (!isValid)
+                {
+                    _tracer.Warning(context, $"SelfCheck: Found invalid entry in cache. Hash={hashInfo.hash}. {error}. Evicting the file...");
+                    await _contentStoreInternal.RemoveInvalidContentAsync(context, hashInfo.hash);
+                    invalidEntries++;
+                }
+
+                // Tracking the progress if needed.
+                traceProgressIfNeeded(hashInfo.hash);
+

```

Comment:
Only call this if the hash has changed from the last hash. Otherwise, we can save that we have processed the hash when we've only processed one of the replicas.